docs/tutorials/developer-liftoff-rust/level-1/1.1-rust-lvl1.mdx

```diff
@@ -0,0 +1,227 @@
+}
+```
+
+If a `query` endpoint must have a prerequisite executed before it can be called, you can define a `guard` function that must be executed before the query function can be executed. If the `guard` function returns an error, the `query` function does not execute.
+
+```rust
+fn guard_function() -> Result<(), String> {
+    // ...
+}
+
+#[ic_cdk::query(guard = "guard_function")]
+fn query_function() {
+    // ...
+}
+```
+
+### `update`
+
+An `update` entry point provides a callable method that can make changes to the canister's state. `update` methods may return a response or they may return an empty type.
+
+
+```rust
+#[ic_cdk::update]
+fn update_function() {
+    // ...
+}
+```
+
+`update` methods can be named or configured to use `guard` functions in the same ways that `query` methods can.
+
+## Candid interface files
+
+Candid is an interface description file (IDL) used to describe a canister's public methods. Each Candid file defines a service, then adds the canister's public methods and the data types that each method accepts and returns.
+
+Here is an example that defines the public method `greet`, which accepts type `text` and returns type `text` using a `query` call.
+
+```did
+service : {
+    "greet" : (text) -> (text) query;
+};
+```
+
+Candid interface description files (`.did`) for Rust canisters must be handwritten or manually exported by first defining the `export_candid!` macro, which exposes a method called `get_candid_pointer`, which `dfx` uses to extract the canister's Candid definition. Then you must use the [`candid-extractor`](https://crates.io/crates/candid-extractor) tool to extract the canister's Candid file. This macro must be called once at the end of your canister's code outside of any query or update definitions.
+
+```rust
+use candid::Principal;
+
+#[ic_cdk::query]
+fn whoami() -> Principal {
+    ic_cdk::caller()
+}
+
+// Export the Candid interface
+ic_cdk::export_candid!();
+```
+
+`query` and `update` methods can be hidden from the Candid exporter by setting `hidden` to equal `true`. If a method is hidden, it still exists in the canister but will not have a Candid interface generated through `export_candid!`
+
+```rust
+#[query(hidden = true)]
+fn query_function() {
+    // ...
+}
+```
+
+## Rust Wasm limitations
+
+WebAssembly (Wasm) is a portable binary format that runs in a virtual machine. ICP compiles canister code to standard Wasm using tools included in the `ic_cdk` that compile Rust code to the `wasm32-unknown-unknown` Wasm target.
+
+Rust relies on the Wasm binary and the associated Candid interface file to define and expose the canister's methods. In a `dfx.json` file that defines a Rust canister, you will need to specify additional fields to indicate where in the project these files are located:
+
+- `build` field: Specifies the command to produce the Wasm binary (using Cargo). If you've defined a package name in the project's `Cargo.toml` file, you can specify the `package` name instead.
```

Comment:
This is a confusing section. It implies that the default, and the primary way you should be thinking about Rust canisters, is `"type": "custom"`, with an optional acceleration called `"type": "rust"` if you have defined a `package` key in `Cargo.toml`. There is no 'if', `package` is a required key, and `"type": "rust"` with `"package"` is the standard form of Rust canister, with a tiny footnote about `"type": "custom"` for bespoke build processes for complex use-cases.